Enable color consistency in a dashboard

khtruong opened the pull request:
We want color consistency so that when users revisit dashboards, colors remain the same for metrics.

Currently, we follow a general algorithm to assign colors to metrics but it can lead to inconsistency. The user first picks a color scheme. Then we iterate through the list of data values and assign a color to it in a sequential manner. The color assignments are stored for the current session. If more values are added, we continue down the list of colors that are part of the scheme. If a user removes a value, all the other values keep their color association because we have stored this information. When the user saves the chart or dashboard, we do not save the color assignments. Thus when the user revisits the chart or dashboard, the color assignments may change because it iterates through the values displayed in the chart or dashboard and re-assigns their colors again. This can cause some confusion when users revisit their visualizations.

This PR enforces color consistency in dashboards so that when users revisit dashboards, the same colors are applied to metrics on each visit. Note that this will not apply until the owner edits his or her dashboard and selects/saves a color scheme.

I am putting out a PR while I add unit tests.

john-bodley committed:
[csv-upload] Fixing message encoding (#6971)

john-bodley committed:
[date-filter] Adding minus infinity (#6968)

* [date-filter] Adding minus infinity

* Update DateFilterControl.jsx

eqeqeq

conglei committed:
Enhancement of query context and object. (#6962)

* added more functionalities for query context and object.

* fixed cache logic

* added default value for groupby

* updated comments and removed print

mistercrunch committed:
auto-set 'Is Temporal' to true where column_name == '__time' (#6857)

* auto-set 'Is Temporal' to true where column_name == '__time'

* Adress comments

* lint

* clarifying name and comments

mistercrunch committed:
[dashboard editor] use the theme's btn-primary (#6859)

khtruong committed:
Sparkline dates aren't formatting in Time Series Table (#6976)

* Exclude venv for python linter to ignore

* Fix NaN error

john-bodley committed:
[sql-parse] Fixing LIMIT exceptions (#6963)

thomaswang-shift committed:
Sort Chart Types based on Usage (#6849)

* initial commit for VIZ-58

* address @kristw's comments

* per @mistercrunch's comment to use a static list for now

* fix javascript test errors

* fix lint errors

* per @betodealmeida's comments

* remove unnecessary loop

mistercrunch committed:
docs: warn that docker installation instructions are `contrib/` (#6925)

* [docs] warn that docker installation instructions are `contrib/`

* Words

michellethomas committed:
Adding custom control overrides (#6956)

* Adding extraOverrides to line chart

* Updating extraOverrides to fit with more cases

* Moving extraOverrides to index.js

* Removing webpack-merge in package.json

* Fixing metrics control clearing metric

kristw committed:
refactor: bump @superset-ui versions  (#6982)

* feat: bump @superset-ui versions

* refactor: remove visUtils and use @superset-ui/dimension

* refactor: remove dead code

* fix: update package lock

* fix: lint two files

* fix: package lock

Commit:
fix(fetch-datasource-meta-data): Chart data is not showing after in charts added by edit mode (#6980)

UIC-1046

xtinec committed:
Split tags migration (#7002)

This PR removes the iteration over charts, dashboards and saved queries to create tags in the original migration, leaving only the logic to create the tags and the tagged objects tables.
Tested locally by running `superset db downgrade` to revert to the previous migration and then running `superset db upgrade` to the current version.

villebro committed:
Retrieve df from get_df_payload to hit cache (#6990)

Commit:
[fix] /superset/slice/id url is too long (#6989)

MarcusSorealheis committed:
fixing a broken link for the ORM documentation (#7005)

mistercrunch committed:
Bump sqlalchemy dependency (#7000)

* Bump sqlalchemy dep

* Addressing comments

michellethomas committed:
Changing time table viz to pass formatTime a date (#7020)

(cherry picked from commit 7f3c145b1f5a4e2d8b95982119503e98772e2c47)

michellethomas committed:
Changing time table viz to pass formatTime a date (#7020)

john-bodley committed:
[db-engine-spec] Aligning Hive/Presto partition logic (#7007)

betodealmeida committed:
Show tooltip with time frame (#6979)

betodealmeida committed:
Fix time filter control (#6978)

thunter009 committed:
[WIP] fix user specified JSON metadata not updating dashboard on refresh (#7027)

conglei committed:
fix inaccurate data calculation with adata rolling and contribution (#7035)

bipinsoniguavus committed:
fix(css-selector): Fixed incorrect css-selector for CRUD page reported by sonarqube report as a critical vulnerability (#7034)

villebro committed:
Add handling for memoryview (#6987)

williaster committed:
[api/v1][query_obj] add default prequeries array (#7039)

* [api/v1][query_obj] add default prequeries array

* [api/v1][query_obj] fix prequeries type

john-bodley committed:
[sqlparse] Fixing table name extraction for ill-defined query (#7029)

Commit:
[fix] Cursor jumping when editing chart and dashboard titles (#7038)

Commit:
fix: groupable only columns populating in filters (#6993)

emacip committed:
User can turn off download - export csv from a dashboard (#6821)

* User can turn off download - export csv from a dashboard

* Revert changes in the backendSync

* Fix space error in the MenuItem Export CSV

Commit:
[fix] explore chart from dashboard missed slice title (#7046)

mistercrunch committed:
Fix issues around Database permissions (#7009)

hajdbo committed:
[fix] Use HTTPS, not HTTP wherever practical (#7040)

* Download RAT binary via HTTPS, not HTTP

* Merge branch 'patch-1' of github.com:hajdbo/incubator-superset into patch-1

john-bodley committed:
[datasource] Ensuring consistent behavior of datasource editing/saving. (#7037)

* Update datasource.py

* Update datasource.py

michellethomas committed:
Adding warning message for sqllab save query (#7028)

kristw committed:
feat: bump superset-ui-plugins to v0.10.2 (#7051)

* feat: bump superset-ui-plugins version to 0.10.1

* fix: bump version again

djok committed:
feat: add Bulgaria to country map (#6938)

mistercrunch committed:
docs: fix RST issues while building docs (#7012)

khtruong committed:
feat: add ability to change font size in big number (#7003)

* Add ability to change font sizes in Big Number

* rename big number to header

* Add comment to clarify font size values

bipinsoniguavus committed:
fix: search for visualizations with case-insensitive keyword

kristw committed:
docs: add hostnfly to company list (#7065)

From old PR #4011

villebro committed:
Handle memoryview like bytes instances and decode bytes to utf8 (#7062)

john-bodley committed:
[tox] Enforcing consistent ordering (#7060)

enricoberti committed:
[SQL Lab] Copy to clipboard should work on the preview tab too (#7068) (#7071)

enricoberti committed:
[SQL Lab] Fixed TableElement sorting functionality and tests (#7069) (#7070)

betodealmeida committed:
Allow LIMIT to be specified in parameters (#7052)

john-bodley committed:
[missing values] Removing replacing missing values (#4905)

mistercrunch committed:
Fix filter_box migration PR #6523 (#7066)

* Fix filter_box migration PR #6523

* Fix druid-related bug

enricoberti committed:
[SQL Lab] Improved query and results tabs rendering reliability (#7082)

closes #7080

khtruong committed:
Enable color consistency in a dashboard

khtruong commented:
👀 @betodealmeida @xtinec @kristw @williaster

khtruong commented:
<img width="1414" alt="Color Consistency in a Dashboard" src="https://user-images.githubusercontent.com/47833996/54791114-ae9b6100-4bf5-11e9-86ca-eb3f4f44f61c.png">

@vylc @dorq

dpgaspar committed:
Bump FAB to 1.12.5 (#7085)

* Bump FAB to 1.12.5

* [requirements] New dependency pyJWT added on FAB 1.12.4 for OAuth redirect

* [requirements] Fix, pip-tools bump to 3.5.0 related to #714

graceguo-supercat commented on superset/assets/src/dashboard/reducers/getInitialState.js:
```diff
@@ -149,6 +157,17 @@ export default function(bootstrapData) {
   };
 
+  // store the color scheme as a layout component so we can undo/redo changes
+  layout[COLOR_SCHEME_ID] = {
```
could you not use `layout`? `layout` stores actual layout in a flattened tree structure. we can traverse the `layout` attribute to build the whole tree. color_scheme should not ride on this specific data structure.

khtruong replied:
Sure! But if I move it out of layout, changing color schemes will not be an undoable/redoable action based on how the things are set up at the moment. If people don't care about it being undoable/redoable, I can definitely move it out.

graceguo-supercat replied:
oh i see, in dashboard metadata it is separated attribute. i am ok with this change. `layout` this name confused me. 👍

khtruong replied:
👍

khtruong replied:
Actually I think I will move it to dashboardState. I explained in reply to your other comment what happens when I do that. What do you think?

graceguo-supercat replied:
if move to dashboardState, remember to add logic to update `hasUnsavedChanges` if user change this state. 

I am ok with both solution, undo-able or not undo-able. we used `dashboardLayout` to represent current editing state. if you feel this name is too close to `layout`, you can re-name to `dashboardEditingState` or `dashboardEditingHistory` or anything, and it can contain all redo-able state besides layout changes.

khtruong replied:
I ended up moving it to dashboardState. I would have liked to leave it in the layout object so it stays undoable but it started to leave to a refactoring effort to rename a bunch of things to provide clarity and reduce confusion. I went with the route that required less unrelated changes and maybe in the future this can be refactored in a separate PR.

graceguo-supercat replied:
NP. Also i should do this rename work.

graceguo-supercat commented:
What is flow look like to change color scheme?
dashboard -> edit mode -> colors button -> change color scheme,
then all the charts in dashboard re-render to reflect the color scheme update, or need to click switch to view mode?

hitdemo committed:
[bugfix] throw "TypeError: 'NoneType'..." when save chart to new/existing dash. (#5835)

khtruong commented:
> What is flow look like when user changes color scheme?
> dashboard -> edit mode -> colors button -> change color scheme,
> then all the charts in dashboard re-render to reflect the color scheme update, or need to click switch to view mode?

The charts will re-render to reflect the color scheme update. No need to switch to view mode.

williaster commented:
wohoo! 🎉 

this is super UI nit but can you offset one of these buttons to avoid the `2px` border divider and instead get a `1px` border 😬

graceguo-supercat commented on superset/assets/src/dashboard/containers/DashboardHeader.jsx:
```diff
@@ -69,14 +69,16 @@ function mapStateToProps({
     ).text,
     expandedSlices: dashboardState.expandedSlices,
     css: dashboardState.css,
+    colorNamespace: undoableLayout.present[COLOR_SCHEME_ID].meta.colorNamespace,
```
for line `layout: undoableLayout.present,` can you remove `COLOR_SCHEME_ID` and `DASHBOARD_HEADER_ID` key from `layout` prop? `layout` prop will be the content to be persisted into dashboard table as `position_json`. COLOR_SCHEME_ID section should not being saved as part of position_json.

khtruong replied:
Thanks for catching that. How about I move the color scheme info to dashboardState? Changing colors will not be an undoable/redoable action if I move it to dashboardState but since the dashboard re-renders and reflects the change immediately, I think it is fine that it is not undoable/redoable.

graceguo-supercat commented:
Superset dashboard used to have label color not consistent issue. 
Suppose we only have 1 color scheme with color [red, green, blue, yellow, pink, black], and dashboard has 3 charts: chart_1 has "boy" and "girl", chart_2 has "cat" and "dog", chart_3 has "coke" and "water".

in old days (before this PR), my understanding that dashboard assigns colors by first-come-first-serve. If query results come back in order [chart_1, chart_2, chart_3], we will see colors are assigned like: 
(boy: red, girl: green), 
(cat:blue, dog:yellow), 
(water: pink, coke: black)

But after a page refresh, the query results may return in another order  [chart_3, chart_1, chart_2], then the color will be assign as
(cat: red, dog: green), 
(water:blue, coke:yellow), 
(boy: pink, girl: black).

In your new solution, can you explain a little bit how you resolve this problem (to make label consistent after page reload)? thanks!

mistercrunch commented:
Really happy to see this!

Does this work in conjunction with http://superset.apache.org/faq.html#is-there-a-way-to-force-the-use-specific-colors ?

This is sometimes needed a user may want to see a success metric in green and errors in red. Maybe eventually the dashboards `Color` section also exposes a color picker for each discovered metrics.

kristw committed:
fix: bump pivot table version (#7088)

* fix: bump pivot table version

* fix: update package.json

khtruong commented:
> wohoo! 🎉
> 
> this is super UI nit but can you offset one of these buttons to avoid the `2px` border divider and instead get a `1px` border 😬
> 
> ![image](https://user-images.githubusercontent.com/4496521/54800836-3c407600-4c21-11e9-81da-a0a1525ef915.png)

Sure thing! I thought it looked a little funky. Good eye 👁

graceguo-supercat commented:
About @mistercrunch comments, adding color picker at label level,  Eli gave me a design at here: https://github.com/apache/incubator-superset/issues/5887. But i don't have time to implement...

khtruong commented:
> Superset dashboard used to have label color not consistent issue.
> Suppose we use color_scheme_1 with colors: [red, green, blue, yellow, pink, black], and dashboard has 3 charts: chart_1 has "boy" and "girl", chart_2 has "cat" and "dog", chart_3 has "coke" and "water".
> 
> in old days (before this PR), my understanding that dashboard assigns colors by first-come-first-serve. If query results come back in order [chart_1, chart_2, chart_3], we will see colors are assigned like:
> (boy: red, girl: green),
> (cat:blue, dog:yellow),
> (water: pink, coke: black)
> 
> But after a page refresh, the query results may return in another order [chart_3, chart_1, chart_2], then the color will be assign as
> (cat: red, dog: green),
> (water:blue, coke:yellow),
> (boy: pink, girl: black).
> 
> In your new solution, can you explain a little bit how you resolve this problem (to make label consistent after page reload)? thanks!

Definitely! Currently when a user loads a dashboard, we check if the user has specified "label_colors" in JSON metadata (http://superset.apache.org/faq.html#is-there-a-way-to-force-the-use-specific-colors). If it is specified, then we take this color mapping. 


i.e. The user specifies the following in JSON metadata:
{
    "label_colors": {
        "cat": "#0000FF",
        "dog": "#FFFF00"
    }
}

We assign these colors to "cat" and "dog" first instead of assigning colors by first-come first-server.


This is the current behavior and is not discoverable unless the user googles how to do this. There is also a subtle bug where this color mapping is associated with the default color scheme "bnbColors".


i.e: When we load a dashboard, the default namespace is "GLOBAL" and the color scheme is ''bnbColors". If "label_colors" is specified, we save this color mapping to the "GLOBAL" namespace and color scheme "bnbColors". If chart_1 uses the color scheme "color_scheme_2", then we will NOT use the "label_colors" color mapping even if they have metrics "cat" and "dog".


With these changes, I leverage "label_colors" fully. When a user saves a color scheme, the color scheme and color mapping is saved to JSON metadata (Namespace can be saved but it is not really leveraged at the moment and is not saved unless it is explicitly specified. But this PR sets it up so that namespace can be easily saved in the future). So when the user revisits the dashboard, we load the color scheme and color mapping. All chart's color scheme is overridden with the dashboard color scheme. So we are always looking up the same color scheme and color mapping.

The only subtly would be if new data values are introduced when users visit the dashboard. If there are new data values, then there are no colors associated with them saved and there can potentially be inconsistency with those values.

khtruong commented:
> Really happy to see this!
> 
> Does this work in conjunction with http://superset.apache.org/faq.html#is-there-a-way-to-force-the-use-specific-colors ?
> 
> This is sometimes needed a user may want to see a success metric in green and errors in red. Maybe eventually the dashboards `Color` section also exposes a color picker for each discovered metrics.

It sure does! When a user loads up an old dashboard, then the experience remains the same. They must go and explicitly choose a color scheme to leverage this feature. If they choose a color scheme, when we save the color scheme and color mapping as "color_scheme" and "label_colors" in the JSON metadata. Now if they choose a color scheme, we may override their previous "label_colors" since the old behavior associated "label_colors" with the color scheme "bnbColors". But they can go and modify this in JSON metadata like before. We also want to allow users to choose their color mapping through a UI but we will tackle that later on. This is the first piece to color consistency.

khtruong commented:
> About @mistercrunch comments, adding color picker at label level, Eli gave me a design at here: #5887. But i don't have time to implement...

I can't wait for this to be implemented! 🙏  I would definitely like the next step to be listing out the data value and color association. Then the third part would be to let users change colors. We can definitely discuss who can tackle this later on. For now, this is the first step to color consistency.

zhaoyongjie committed:
[Dashboard] Keeping refresh frequency value in Dashboard (#5741)

* Keeping refresh frequency value in Dashboard

* fix eslint

* Fix arrow function

* Update ut

* start refresh when load dashboard

khtruong committed:
Moved actions, minor UI, allowed dashboard copy

williaster committed:
[api] add `available_domains/` endpoint (#7097)

* [api] add available_domains/ endpoint

* [available_domains] no trailing whitespaces

Commit:
handle null column_name in sqla and druid models (#7063)

john-bodley committed:
[mypy] Adding mypy linting (#7053)

villebro committed:
Bump SQLAlchemy to 1.3 (#7099)

* Bump sqla to >=1.3.1

* Refine mssql column types to only use N-prefixing when necessary

* make join explicit

* replace set with list

* Add additional test case for N-prefix

* Replace engine with dialect and fix linting error

* Remove unneeded import

john-bodley committed:
[forms] Fix handling of NULLs (#7078)

khtruong committed:
Fix linting errors

khtruong committed:
Undo unintentional change

michellethomas committed:
Use metric name instead of metric in filter box (#7106)

khtruong committed:
Updated and added unit tests

khtruong committed:
Fail quietly if package has not been updated

mistercrunch committed:
[docs] improve UPDATING.md around 0.31's [now] optional deps (#7117)

xtinec committed:
SQL editor layout makeover (#7102)

This PR includes the following layout and css tweaks:
- Using flex to layout the north and south sub panes of query pane so resizing works properly in both Chrome and Firefox
- Removal of necessary wrapper divs and tweaking of css in sql lab so we can scroll to the bottom of both the table list and the results pane
- Make sql lab's content not overflow vertically and layout the query result area to eliminate double scroll bars
- css tweaks on the basic.html page so the loading animation appears in the center of the page across the board

khtruong committed:
Fail quietly on dashboard copy if package is old

donbowman committed:
Add HTTP Basic Auth for Druid (Issue #4776) (#6795)

* Add HTTP Basic Auth for Druid (Issue #4776)

This should resolve [Issue #4776](https://github.com/apache/incubator-superset/issues/4776)

Add HTTP Basic Auth to the Druid Broker.

Signed-off-by: Don Bowman <don@agilicus.com>

* Encrypt broker password in database

Signed-off-by: Don Bowman <don@agilicus.com>

* Merge from master, new migration needed

Signed-off-by: Don Bowman <don@agilicus.com>

* Do not expose hashed druid basic auth password

* docs: add description_columns for druid basic auth

mistercrunch committed:
Scripts & instructions to craft official Apache source releases (#7054)

* Apache Releases

https://github.com/mistercrunch/superset/blob/apache-releases/RELEASING.md#apache-releases

* npm run prod -> npm run build

rasmi-ranjan-guavus committed:
fix(useless-self-assignment): correct useless-self-assignment (#7090)

* fix(useless-self-assignment): correct useless-self-assignment

* udated code formatting

mistercrunch committed:
Remove super verbose `logging.debug` call (#7122)

I left this behind on a previous PR, and it gets kind of crazy on large
SQL statements, so decided it's probably a good thing to clean this up.

mistercrunch committed:
Bump python lib croniter to an existing version (#7132)

Package maintainers should really never delete packages, but it appears
this happened with croniter and resulted in breaking our builds.

This PR bumps to a more recent existing version of the library

khtruong commented:
Closing because I got my branch in a weird state and it was just easier to create a new branch.

Refer to this PR: https://github.com/apache/incubator-superset/pull/7135
